runtime/shared_common/shrinit.cpp

```diff
@@ -4923,4 +4931,201 @@ isFreeDiskSpaceLow(J9JavaVM *vm, U_64* maxsize)
 	}
 	return ret;
 }
+
+/**
+ * This function generates a key for the GC hints. The key is a string of all the commandline arguments passed to the current JVM
+ * @param[in] vm The current J9JavaVM
+ *
+ * @return A string of all the commandline arguments, NULL if an error occurs.
+ */
+static char*
+generateStartupHintsKey(J9JavaVM* vm)
+{
+	JavaVMInitArgs *actualArgs = vm->vmArgsArray->actualVMArgs;
+	UDATA keyLength = 0;
+	UDATA extraLen = 0;
+	UDATA i = 0;
+	UDATA nOptions = vm->vmArgsArray->nOptions;
+	char* key = NULL;
+	bool firstOption = true;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	for (i = 0; i < nOptions; ++i) {
+		char* option = actualArgs->options[i].optionString;
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"-Dsun.java.launcher.pid="))
+		) {
+			keyLength += strlen(actualArgs->options[i].optionString);
+			extraLen += 1;
+		}
+	}
+	if (keyLength == 0) {
+		goto done;
+	}
+	keyLength += extraLen; /* space between options and terminating null char */
+
+	key = (char*)j9mem_allocate_memory(keyLength, J9MEM_CATEGORY_VM);
+	if (NULL == key) {
+		goto done;
+	}
+	memset(key, 0, keyLength);
+	for (i = 0; i < nOptions; ++i) {
+		char* option = actualArgs->options[i].optionString;
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"sun.java.launcher.pid"))
+		) {
+			if (firstOption) {
+				firstOption = false;
+				j9str_printf(PORTLIB, key, keyLength, "%s%s", key, option);
+			} else {
+				j9str_printf(PORTLIB, key, keyLength, "%s%s%s", key, " ", option);
+			}
+		}
+	}
+done:
+	return key;
+}
+
+
+/**
+ * This function fetches the startup hints from the shared cache and store it locally to vm->sharedClassConfig->localStartupHints.
+ * @param[in] currentThread  The current VM thread
+ *
+ */
+static void
+fetchStartupHintsFromSharedCache(J9VMThread* currentThread)
+{
+	J9JavaVM* vm = currentThread->javaVM;
+
+	if (NULL != vm->sharedClassConfig) {
+		char *key = NULL;
+		if (NULL != vm->sharedClassConfig->localStartupHints.key) {
```

Comment:
I'm suggesting to free the key, maybe this should use a flag to determine if the hints have been fetched.

Reply:
Removed key from struct J9SharedLocalStartupHints and used a flag to determine if hints are fetched.